### ISSUE TITLE
Kivételkezelés: logolás és csomagolás

### ISSUE THREAD

**Issue by tmichel** [kir-dev/korok#41]
Át kell nézni a kivételkezelést az EJB rétegben. Minden kivétel esetén el kell dönteni, hogy helyben kezeljük-e le, vagy inkább tovább dobjuk a web rétegbe.

A kivételek logolását mindenhol be kell vezetni. Esetleg a logolási szintet lehetne egyeztetni és valami konvenciót megfogalmazni az utókor számára.

A webrétegbe csak a saját kivételeinkbe csomagolt system exceptionöket dobjuk tovább. Ne kerüljön ki "nyers" kivétel az ejb rétegből.

Internal api miatt módosul a klasszikus webrétegbeli hibakezelés. Maradunk annál a koncepciónál, hogy egyetlen alkalmazás szintű kivételt dobálunk át a web rétegbe, de itt már olyan információkat kell visszaadnunk a kliensnek, amivel a felhasználót tudja értesíteni. 

Pl ha egy unique indexet sértene egy módosítás, akkor ideális esetben a hibajelentés tartalmazná, hogy melyik mezővel milyen baj volt.

Ennek a segítésére a `PekException` osztályt fel kellene okosítani annyira, hogy kontextus függő információt is át lehessen adni és mindent valami könnyen feldolgozható formában. (kulcs-érték párok?)

**Comment by vbalazs** [comment]
Át kell nézni a sample logback.xml konfigurációt is, nem egészen világos melyik log melyik fájlba kerül is miért.
Ki kell találni rendesen és megcsinálni hozzá a logback configot.

Én ott választanám külön, hogy web | minden más. Bármilyen ötlet / best practice jöhet.

**Comment by tmichel** [comment]
A kivételkezeléshez létrehozunk néhány (minél kevesebb) `ApplicationExceptiont`. Az egyes hibatípusok megkülönböztetésére hibakódokat használunk.

A hibakódokat `enum`-ban írjuk le. Tartozik majd hozzájuk egy property fájlba felvehető string kulcs is, így a felhasználói értesítés is megoldható a webrétegben viszonylag fájdalommentesen.

Az exceptionök részletei kidolgozásra várnak, ahogy a hibakódok is. Szerintem a hibakódoknál érdemes lenne a HTTP status code-k struktúrájához hasonlóban gondolkodni. 1xx jelenet valamint, 2xx jelent valamit stb.

**Comment by messo** [comment]
A hibatípusok - hibakódok <-> exception rész nem világos. Minek az enum? Mindegyik esetre külön exception, amit az `ApplicationException`-ből származtattok, vagy ti olyat csinálnátok, hogy `throw new ApplicationException(ErrorCodes.DAFAQ_JUST_HAPPENED)` ?

**Comment by vbalazs** [comment]
> A hibatípusok - hibakódok <-> exception rész nem világos. Minek az enum? Mindegyik esetre külön exception, amit az ApplicationException-ből származtattok, vagy ti olyat csinálnátok, hogy throw new ApplicationException(ErrorCodes.DAFAQ_JUST_HAPPENED) ?

utóbbira gondoltunk :panda_face:

**Comment by tmichel** [comment]
@messo pontosan mint a második, amit írsz. többek között az szeretnénk elkerülni hogy a webben úgy nézzek ki a dolog, hogy

``` java
try {
    someOperation();
} catch (...) {
} catch (...) {
} catch (...) {
} catch (...) {
} catch (...) {
}
```

Most hirtelen nagyon efelé indultunk és ez így nem túl karbantartható.

**Comment by messo** [comment]
Hmm, erre egy ötletesebb exception hierarchia vagy JDK7-es `catch (AException | BException | CException ex)` lehet a megoldás?

**Comment by vbalazs** [comment]
használtuk pl itt: https://github.com/kir-dev/korok/commit/927615f63ec2df1589f6ff4e3d0b0cb8bd3a2179#L14R159

de ez sem tökéletes megoldás.
ha az exceptionoket nagyon hierarchiába rendezzük és úgy kapkodjuk el a szülő exceptionként (pl: InvalidNewbieStateException < RegistrationException, UserAlreadyExistsException < RegistrationException) akkor meg általános exceptionbol lesz megint para, pl. UserNotFoundException.

azért jó az általános exception használata, mert általánosít (ki gondolta volna?:D) tulajdonképpen nekunk ritkán kell a kódban különböző műveleteket végezni különböző exception esetén (a webrétegről van szó!), inkább a hibaüzenet kijelzését végezzük csak. Ehhez meg felesleges checked exceptionoket gyártani imho.

**Comment by messo** [comment]
Stimmt, ti látjátok mit szeretnétek, ha tényleg nem kell semmilyen elágazó logika, akkor fincsi a paraméterezett exception. Ahol pedig esetleg kellhet ott lehet érdemesebb inkább két kül. exception, minthogy az enum értékétől függően csináltok valamit.

**Comment by vbalazs** [comment]
írtam egy ilyet:
https://redmine.kirdev.sch.bme.hu/projects/vip/wiki/PEKKonvenciok#Naplzs

**Comment by tmichel** [comment]
frissítettem a leírást az apival együtt jelentkező új igényekkel
